crates/compilers/src/preprocessor/data.rs

```diff
@@ -0,0 +1,214 @@
+            constructor_data,
+            artifact,
+        }
+    }
+
+    /// If contract has a non-empty constructor, generates a helper source file for it containing a
+    /// helper to encode constructor arguments.
+    ///
+    /// This is needed because current preprocessing wraps the arguments, leaving them unchanged.
+    /// This allows us to handle nested new expressions correctly. However, this requires us to have
+    /// a way to wrap both named and unnamed arguments. i.e you can't do abi.encode({arg: val}).
+    ///
+    /// This function produces a helper struct + a helper function to encode the arguments. The
+    /// struct is defined in scope of an abstract contract inheriting the contract containing the
+    /// constructor. This is done as a hack to allow us to inherit the same scope of definitions.
+    ///
+    /// The resulted helper looks like this:
+    /// ```solidity
+    /// import "lib/openzeppelin-contracts/contracts/token/ERC20.sol";
+    ///
+    /// abstract contract DeployHelper335 is ERC20 {
+    ///     struct ConstructorArgs {
+    ///         string name;
+    ///         string symbol;
+    ///     }
+    /// }
+    ///
+    /// function encodeArgs335(DeployHelper335.ConstructorArgs memory args) pure returns (bytes memory) {
+    ///     return abi.encode(args.name, args.symbol);
+    /// }
+    /// ```
+    ///
+    /// Example usage:
+    /// ```solidity
+    /// new ERC20(name, symbol)
+    /// ```
+    /// becomes
+    /// ```solidity
+    /// vm.deployCode("artifact path", encodeArgs335(DeployHelper335.ConstructorArgs(name, symbol)))
+    /// ```
+    /// With named arguments:
+    /// ```solidity
+    /// new ERC20({name: name, symbol: symbol})
+    /// ```
+    /// becomes
+    /// ```solidity
+    /// vm.deployCode("artifact path", encodeArgs335(DeployHelper335.ConstructorArgs({name: name, symbol: symbol})))
+    /// ```
+    pub fn build_helper(&self) -> Option<String> {
+        let Self { contract_id, path, name, constructor_data, artifact } = self;
+
+        let Some(constructor_details) = constructor_data else { return None };
+        let contract_id = contract_id.get();
+        let struct_fields = &constructor_details.struct_fields;
+        let abi_encode_args = &constructor_details.abi_encode_args;
+        let vm_interface_name = format!("VmContractHelper{contract_id}");
+        let vm = format!("{vm_interface_name}(0x7109709ECfa91a80626fF3989D68f67F5b1DD12D)");
+
+        let helper = format!(
+            r#"
+pragma solidity >=0.4.0;
+
+import "{path}";
+
+abstract contract DeployHelper{contract_id} is {name} {{
+    struct ConstructorArgs {{
+        {struct_fields};
+    }}
+}}
+
+function encodeArgs{contract_id}(DeployHelper{contract_id}.ConstructorArgs memory args) pure returns (bytes memory) {{
+    return abi.encode({abi_encode_args});
+}}
+
+function deployCode{contract_id}(DeployHelper{contract_id}.ConstructorArgs memory args) returns({name}) {{
```

Comment:
I think we can replace this helper function with just a call to `vm.deployCode(<object>)` and we programmatically build up `object` with named arguments like `push(("artifact", "...")); if arguments { push(("data", "encodeArgs({...})")) }` etc

Reply:
:+1: added with https://github.com/foundry-rs/compilers/pull/252/commits/6a8114749d040997390cef5e5483f2a40b791bc8 requires https://github.com/foundry-rs/foundry/pull/10178 to support all cheatcodes

Reply:
new cheatcodes were merged, I added foundry test for constructors with value and salt in https://github.com/foundry-rs/foundry/pull/10010/commits/4c6939656a6ae46b5c689db8479fe85f7704e352